File: JavaScriptPractice/Leetcode54.js
var spiralOrder = function(matrix) 
{
    let m =matrix.length;
    let n= matrix[0].length;
    let step = m*n;
    let directions = [[0,1],[1,0],[0,-1],[-1,0]]
    let directIndex = 0
    let curRow =0;
    let curCol = 0;
    let visited = Array.from({length:m},()=>Array(n).fill(false))
    let res = []
    while (step>0) {
        res.push(matrix[curRow][curCol])
        visited[curRow][curCol]=true
        step--
        let nextRow = curRow+ directions[directIndex][0];
        let nextCol = curCol+ directions[directIndex][1];
        if (nextRow<0||nextCol<0||nextRow>m-1||nextCol>n-1||visited[nextRow][nextCol]==true) {
            directIndex =(directIndex+1)%4
            
        }
        curRow = curRow+ directions[directIndex][0];
        curCol = curCol+ directions[directIndex][1];
    }
    return res
};

File: CppPractice/CppLeetcode/Leetcode54.cpp
#include <vector>
using namespace std;
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) 
    {
        vector<int> res;
        int rowNum = matrix.size();
        int colNum = matrix[0].size();
        int step = rowNum*colNum; 

        vector<vector<bool>> visited(m,vector<bool>(n,false));

        int curRow = 0;
        int curCol = 0;
        int directIndex = 0;
        vector<pair<int,int>> directions;
        directions.push_back(make_pair(0,1));
        directions.push_back(make_pair(1,0));
        directions.push_back(make_pair(0,-1));
        directions.push_back(make_pair(-1,0));

        while (step>0)
        {
            res.push_back(matrix[curRow][curCol]);
            step--;

            int nextRow = curRow + directions[directIndex].first;
            int nextCol = curRow + directions[directIndex].second;
            if(nextRow<0||nextCol<0||nextRow>m-1||nextCol>n-1||visited[nextRow][nextCol]==true)
            {
                directIndex = (directIndex+1)%4;
                cur = curRow + directions[directIndex].first;
                cur = curRow + directions[directIndex].second;
            }

        }
        return res;
    }
};